### PR TITLE
fix: test for sales order, pick list and sales invoice

### PR DESCRIPTION
**Pull request by imdadhussain**
**Task ID**:  https://bloomstack.com/desk#Form/Task/TASK-2021-00345
 
- [x] Fixes

**Description**: 

1. In Sales Order Test cases there is Permission error validation
2. In Sales Order In items Details object set None value.
3. In Pick List Test cases warehouses name error.
4. In Sales Invoice Test Cases Some are Item Code related Issue.
5.  In AccountingDimension  Test Cases ValidationError Raise error.
6. In Invoice Discounting Test Cases Some are Item Code related Issue.

**Tested**: YES 

List of Test cases Solved:
```
- test_request_for_raw_materials (erpnext.selling.doctype.sales_order.test_sales_order.TestSalesOrder)
- test_warehouse_user (erpnext.selling.doctype.sales_order.test_sales_order.TestSalesOrder)
-  test_update_child_qty_rate_perm (erpnext.selling.doctype.sales_order.test_sales_order.TestSalesOrder)
- test_serial_no_based_delivery (erpnext.selling.doctype.sales_order.test_sales_order.TestSalesOrder)
- test_pick_list_for_items_from_multiple_sales_orders (erpnext.stock.doctype.pick_list.test_pick_list.TestPickList)
- test_pick_list_picks_warehouse_for_each_item (erpnext.stock.doctype.pick_list.test_pick_list.TestPickList)
- test_pick_list_shows_serial_no_for_serialized_item (erpnext.stock.doctype.pick_list.test_pick_list.TestPickList)
- test_pp_to_mr_customer_provided (erpnext.manufacturing.doctype.production_plan.test_production_plan.TestProductionPlan)
- test_production_plan (erpnext.manufacturing.doctype.production_plan.test_production_plan.TestProductionPlan)
- test_production_plan_for_existing_ordered_qty (erpnext.manufacturing.doctype.production_plan.test_production_plan.TestProductionPlan) 
- test_production_plan_with_non_stock_item (erpnext.manufacturing.doctype.production_plan.test_production_plan.TestProductionPlan)
- test_production_plan_without_multi_level (erpnext.manufacturing.doctype.production_plan.test_production_plan.TestProductionPlan)
- test_production_plan_without_multi_level_for_existing_ordered_qty (erpnext.manufacturing.doctype.production_plan.test_production_plan.TestProductionPlan)
- test_create_so_with_margin (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_return_sales_invoice (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice) 
- test_rounding_adjustment_2 (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_sales_invoice_for_disable_allow_cost_center_in_entry_of_bs_account (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_sales_invoice_for_enable_allow_cost_center_in_entry_of_bs_account (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_sales_invoice_gl_entry_with_perpetual_inventory_no_item_code (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_sales_invoice_gl_entry_with_perpetual_inventory_non_stock_item (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_sales_invoice_gl_entry_without_perpetual_inventory (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_sales_invoice_with_advance (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_serialized (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_serialized_cancel (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_sales_invoice_with_discount_and_inclusive_tax (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_serial_numbers_against_delivery_note (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
- test_make_payment_before_after_period (erpnext.accounts.doctype.invoice_discounting.test_invoice_discounting.TestInvoiceDiscounting)
- test_make_payment_before_loan_period (erpnext.accounts.doctype.invoice_discounting.test_invoice_discounting.TestInvoiceDiscounting)
- test_mandatory (erpnext.accounts.doctype.accounting_dimension.test_accounting_dimension.TestAccountingDimension)
```
**Trackback/Screenshot:**: 
        **Sales Order  Test**
Screenshot 2021-10-01 at 9 55 13 AM

Screenshot 2021-10-01 at 11 47 23 AM

        **Pick List  Test**
Screenshot 2021-10-01 at 9 54 54 AM

 **Sales Invoice  Test Cases**

1. SI-01

2. SI-02

3. SI-03

4. SI-04

5. SI-05

6. SI-06
 
7. SI-07

8. SI-08

 **Invoice Discounting**

ID-1

 **Accounting Dimension**

AD-1

### CODE REVIEW

**Comment by vishdha** [comment]
@imdad can you add traceback in the description

File: erpnext/stock/doctype/serial_no/serial_no.py [hunk 1]
```diff
@@ -431,7 +431,6 @@ def update_args_for_serial_no(serial_no_doc, serial_no, args, is_new=False):
 		"via_stock_ledger": args.get("via_stock_ledger") or True,
 		"supplier": args.get("supplier"),
 		"location": args.get("location"),
-		"sales_order": frappe.db.get_value(args.get("voucher_type"), args.get("voucher_no"), "sales_order_no"),
```
Comment (sahil28297) [inline erpnext/stock/doctype/serial_no/serial_no.py]: @imdadhussain remove this file change, track it from https://github.com/Bloomstack/erpnext/pull/1742

**Review by sahil28297 — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/serial_no/serial_no.py (hunk 1 above)
Comment (imdadhussain) [reply erpnext/stock/doctype/serial_no/serial_no.py]: @sahil28297 it is merge on develop so that we can pull it on local.

**Comment by Anurag810** [comment]
Please update the description as per the template. it should be like a non-functional guy can also undestand.

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 2]
```diff
@@ -564,7 +564,8 @@ def get_subitems(doc, data, item_details, bom_no, company, include_non_stock_ite
 				if d.qty > 0:
 					get_subitems(doc, data, item_details, d.default_bom, company,
 						include_non_stock_items, include_subcontracted_items, d.qty)
-		item_details['sales_order'] = sales_order
+		if sales_order:
+			item_details['sales_order'] = sales_order
```
Comment (sahil28297) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py:568]:
```suggestion
		item_details['sales_order'] = sales_order if sales_order else ""
```

File: erpnext/selling/doctype/sales_order/test_sales_order.py [hunk 3]
```diff
@@ -423,7 +423,7 @@ def test_warehouse_user(self):
 
 		test_user = frappe.get_doc("User", "test@example.com")
 		test_user.add_roles("Sales User", "Stock User")
-		test_user.remove_roles("Sales Manager")
+		test_user.remove_roles("Sales Manager", "System Manager")
```
Comment (sahil28297) [inline erpnext/selling/doctype/sales_order/test_sales_order.py:426]: why are you removing the system manager role?

**Review by sahil28297 — CHANGES_REQUESTED** [review]

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 4]
```diff
@@ -564,7 +564,7 @@ def get_subitems(doc, data, item_details, bom_no, company, include_non_stock_ite
 				if d.qty > 0:
 					get_subitems(doc, data, item_details, d.default_bom, company,
 						include_non_stock_items, include_subcontracted_items, d.qty)
-		item_details['sales_order'] = sales_order
+		item_details['sales_order'] = sales_order if sales_order else ""
```
Comment (sahil28297) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]:
```suggestion
		if sales_order:
			item_details['sales_order'] = sales_order
```

**Review by sahil28297 — CHANGES_REQUESTED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**